Build separate full contributors list

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

In https://github.com/WordPress/gutenberg/issues/38749 we learned of a few reasons why listing the usernames of users next to their respective PRs might not be a great idea.

This PR changes things up so we have a single list of contributors underneath a `Contributors` section. It looks like this for `12.5`

```
## Contributors

The following users merged PRs in this release:

@aaronrobertshaw @adamziel @ajlende @alanjacobmathew @alexstine @amustaque97 @andrewserong @annezazu @anton-vlasenko @arthur791004 @audrasjb @c4rl0sbr4v0 @carolinan @chad1008 @ciampo @cr0ybot @dcalhoun @dmsnell @fabiankaegy @fluiddot @fullofcaffeine @geriux @getdave @glendaviesnz @guarani @gziolo @hypest @imangm @jameskoster @jasmussen @jeremyfelt @jhnstn @johnbillion @jorgecontreras @jorgefilipecosta @kebbet @kevin940726 @KevinBatdorf @knilkantha @Mamaduka @mchowning @mkevins @ndiego @noisysocks @ntsekouras @oandregal @ocean90 @ockham @peterwilsoncc @ramonjd @ribaricplusplus @richtabor @ryelle @scruffian @SiobhyB @talldan @tellthemachines @vcanales @walbo @youknowriad
```

**Note**: I'm not sure if this will cause GH to auto-generate it's own list thus duplicating the list. @priethor this is one of the reasons why I avoided this approach in my previous PR.

Closes https://github.com/WordPress/gutenberg/issues/38749

## Testing Instructions

- Run `npm run changelog -- --milestone="Gutenberg 12.6` in Gutenberg repo root.
- See generated changelog.
- `First time contributors` should be "as was"
- Check `Contributors` section is there and includes a single entry per username.
- Check not "bot" users are listed in either `First time contributors` or `Contributors` sections.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

getdave committed:
Build separate full contributors list

priethor reviewed (commented):
Thanks for iterating on this, Dave!

Should we add an array of exclusions to avoid listing bots like `@dependabot[bot]`?

getdave commented:
> Thanks for iterating on this, Dave!
> 
> Should we add an array of exclusions to avoid listing bots like `@dependabot[bot]`?

We should. If we're happy with this format then I'll refine this PR to do that cleanup.

priethor commented:
The format works quite well, in my opinion. If after a few releases we see it doesn't, we can iterate iterate on it 👍

getdave committed:
Add universal function to filter bot users and apply to both pipelines

getdave committed:
Add bot PR to fixture data.

getdave committed:
Update snapshots

getdave committed:
Fix JSON error in fixture data

getdave committed:
Ensure bot users are excluded from the changelog

getdave committed:
Add unit test for skipUsers

getdave committed:
Add snapshot test for contributors list

getdave commented on bin/plugin/commands/changelog.js:
```diff
@@ -683,7 +681,10 @@ async function fetchAllPullRequests( octokit, settings ) {
 
-	const groupedPullRequests = groupBy( pullRequests, getIssueType );
+	const groupedPullRequests = groupBy(
+		skipUsers( pullRequests ),
```
Note @priethor we are now skipping "bots" from the main changelog output itself. Do we want to do that?

gziolo replied:
Yes, that makes a lot of sense because we have now dependabot polluting the changelog 😅

priethor replied:
Yes, we should do that 👍

getdave commented on bin/plugin/commands/changelog.js:
```diff
@@ -781,10 +782,6 @@ function sortFeatureGroups( featureGroups ) {
  */
 function getFirstTimeContributorPRs( pullRequests ) {
 	return pullRequests.filter( ( pr ) => {
-		if ( pr.user.type.toLowerCase() === 'bot' ) {
```
Now handled in dedicated `skipUsers` function.

getdave commented on bin/plugin/commands/changelog.js:
```diff
@@ -822,7 +819,7 @@ function getContributorMarkdownList( ftcPRs ) {
  */
 function sortByUsername( items ) {
-	return sortBy( items, ( item ) => item.user.login );
+	return sortBy( items, ( item ) => item.user.login.toLowerCase() );
```
Sorting by lowercase required else those with capitals in usernames end up _before_ those which come alphabetically before them.

getdave commented on bin/plugin/commands/test/fixtures/pull-requests.json:
```diff
@@ -15265,5 +15265,112 @@
 		},
 		"body": "<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->\r\n\r\n## Description\r\n<!-- Please describe what you have changed or added -->\r\n\r\nRelated: #22890 his PR refactors the `MetaBoxesArea` component into a function component using React hooks.\r\n\r\n##  ...
 		"performed_via_github_app": null
+	},
+	{
+		"url": "https://api.github.com/repos/WordPress/gutenberg/issues/XXXX",
+		"repository_url": "https://api.github.com/repos/WordPress/gutenberg",
+		"labels_url": "https://api.github.com/repos/WordPress/gutenberg/issues/XXXX/labels{/name}",
+		"comments_url": "https://api.github.com/repos/WordPress/gutenberg/issues/XXXX/comments",
+		"events_url": "https://api.github.com/repos/WordPress/gutenberg/issues/XXXX/events",
+		"html_url": "https://github.com/WordPress/gutenberg/pull/XXXX",
+		"id": 123456,
+		"node_id": "123456",
+		"number": 123456,
+		"title": "Placeholder to test that PRs by bots are removed.",
+		"user": {
+			"login": "@dependabot[bot]",
+			"id": 123456,
+			"node_id": "123456",
+			"avatar_url": "https://avatars.githubusercontent.com/u/123456?v=4",
+			"gravatar_id": "",
+			"url": "https://api.github.com/users/torounit",
+			"html_url": "https://github.com/torounit",
+			"followers_url": "https://api.github.com/users/torounit/followers",
+			"following_url": "https://api.github.com/users/torounit/following{/other_user}",
+			"gists_url": "https://api.github.com/users/torounit/gists{/gist_id}",
+			"starred_url": "https://api.github.com/users/torounit/starred{/owner}{/repo}",
+			"subscriptions_url": "https://api.github.com/users/torounit/subscriptions",
+			"organizations_url": "https://api.github.com/users/torounit/orgs",
+			"repos_url": "https://api.github.com/users/torounit/repos",
+			"events_url": "https://api.github.com/users/torounit/events{/privacy}",
+			"received_events_url": "https://api.github.com/users/torounit/received_events",
+			"type": "bot",
```
I've manually added this entry to the fixture data to simulate a PR by a "bot" user. This allows us to verify it is excluded in the final output.

gziolo replied:
Maybe we should keep it in a separate file and merge with the one autogenerated from an old release? The risk is that we would lose this custom entry if it gets regenerated in the future. It’s not a blocker but it’s something that we won’t have under control moving forward.

getdave replied:
✅ Agreed. Done.

getdave committed:
Tidy snapshots

bph reviewed (approved):
This is perfect for the contributors who delivered code!  @getdave :-) 
Not sure if that has any bearing on this PR. When I used the command in your test instructions, I got back the error message: "Cannot find milestone by title Gutenberg 12.5". I got the list when I switched to 12.6 so no problem. 

Which got me thinking, if that list is only available as long as there is not new release, we might need to consider storing it some where. If that's the Make Release block or the release notes on GitHub that's fine, we can copy/paste it into an overall list for the WordPress release. 

Awesome.

getdave commented:
Thanks for your review @bph 🙇‍♂️ 

> Which got me thinking, if that list is only available as long as there is not new release

Once this PR is merged the next release (`12.7` maybe?) will show the list on the [`Releases` tab in Github](https://github.com/WordPress/gutenberg/releases) under the appropriate release.

This will then continue for all future releases meaning it will never be lost.

gziolo commented on bin/plugin/commands/changelog.js:
```diff
@@ -836,19 +833,32 @@ function getUniqueByUsername( items ) {
+ * @param {IssuesListForRepoResponseItem[]} pullRequests List of pull requests.
+ * @return {IssuesListForRepoResponseItem[]} The list of filtered pull requests.
+ */
+function skipUsers( pullRequests ) {
```
Nit: Maybe `skipCreatedByBots` for now? It was a bit unclear what it does in other places.

priethor replied:
The `skipUsers` function was inspired by #38538. Probably a rename would help clarify its purpose, such as `skipCreatedByBots` or `skipFakeContributors`.

getdave replied:
Yeh actually for now renaming to include the "bots" term would be helpful. I was originally thinking it might be a catch all solution but there isn't a use case for that at the moment so let's keep things simple.

getdave replied:
✅

gziolo commented on bin/plugin/commands/changelog.js:
```diff
@@ -860,6 +870,46 @@ function getContributors( pullRequests ) {
+			markdownList += ` @${ pr.user.login }`;
+			return markdownList;
+		}, '' )
+		.trim();
```
Nice trick with trimming the output 👍

getdave replied:
🥳

gziolo reviewed (approved):
Thank you for another iteration. Let’s give it a go and see how it looks on the release page. If there is too much duplication with the list of contributors added by GitHub we can always try to wrap the section in HTML comments and see if that helps.

I left some minor implementation notes to consider but not blockers.

getdave committed:
Rename skip function for clarity of purpose

Addresses https://github.com/WordPress/gutenberg/pull/38777#discussion_r808986091

getdave committed:
Separate bot fixture for greater control

getdave commented:
Will merge once all the tests pass.
